### PR TITLE
planner, sessionctx: skip wide type columns such as JSON when collecting statistics

### LINKED ISSUE #44725: Skip collecting statistics for some wide column types such as JSON/TEXT/BLOB
## Enhancement

Currenly, analyze would collect about 11000 samples into TiDB memory. If the table has some wide column types such JSON/TEXT/BLOB, the samples can take up much memory and even lead to OOM. However, the statistics collected for JSON/BLOB are not very important since the optimizer rarely use them. Hence, we hope to skip collecting statistics for some wide column types such as JSON/TEXT/BLOB.

### PR DESCRIPTION
**Pull request by xuyifangreeneyes**
### What problem does this PR solve?

Issue Number: close #44725

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -2467,6 +2467,24 @@ func (b *PlanBuilder) buildAnalyzeFullSamplingTask(
 		if colsInfo, ok := colsInfoMap[physicalID]; ok {
 			execColsInfo = colsInfo
 		}
+		filterSkipColumnTypes := func(origin []*model.ColumnInfo) (result []*model.ColumnInfo) {
+			skipTypes := b.ctx.GetSessionVars().AnalyzeSkipColumnTypes
+			if b.ctx.GetSessionVars().InRestrictedSQL {
+				val, err1 := b.ctx.GetSessionVars().GlobalVarsAccessor.GetGlobalSysVar(variable.TiDBAnalyzeSkipColumnTypes)
```
Comment (chrysan) [inline planner/core/planbuilder.go:2473]: It will not impact the column list persisted in analyze_options, right?

Comment (xuyifangreeneyes) [reply planner/core/planbuilder.go:2473]: Yes.

**Review by chrysan — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

File: planner/core/planbuilder.go [hunk 2]
```diff
@@ -2467,6 +2467,24 @@ func (b *PlanBuilder) buildAnalyzeFullSamplingTask(
 		if colsInfo, ok := colsInfoMap[physicalID]; ok {
 			execColsInfo = colsInfo
 		}
+		filterSkipColumnTypes := func(origin []*model.ColumnInfo) (result []*model.ColumnInfo) {
+			skipTypes := b.ctx.GetSessionVars().AnalyzeSkipColumnTypes
```
Comment (qw4990) [inline planner/core/planbuilder.go:2471]: The background auto-analyze goroutine's `sctx` is initialized at the beginning, so can it read the latest variable value here?

**Comment by qw4990** [comment]
Should we pick this PR back to LTS7.1 and LTS6.5? This feature is very useful.

File: sessionctx/variable/varsutil.go [hunk 3]
```diff
@@ -574,3 +574,41 @@ var GAFunction4ExpressionIndex = map[string]struct{}{
...
+	"text":       {},
+	"mediumtext": {},
+	"longtext":   {},
+	"blob":       {},
+	"mediumblob": {},
+	"longblob":   {},
+}
+
+// ValidAnalyzeSkipColumnTypes makes validation for tidb_analyze_skip_column_types.
+func ValidAnalyzeSkipColumnTypes(val string) (string, error) {
```
Comment (qw4990) [inline sessionctx/variable/varsutil.go:589]: Why do we add this validation? It seems that all types should be allowed to skip.

Comment (xuyifangreeneyes) [reply sessionctx/variable/varsutil.go:589]: For caution, we constrain the column types which can be skipped to `json,blob,mediumblob,longblob,text,mediumntext,longtext`.

Comment (xuyifangreeneyes) [reply sessionctx/variable/varsutil.go:589]: Skipping a int-type column doesn't have much benefit but accidentally `set tidb_analyze_skip_column_types = 'int'` may bring some trouble.

**Review by qw4990 — APPROVED** [review]

**Comment by qw4990** [comment]
Better to add more tests.

**Review by qw4990 — APPROVED** [review]

**Review by chrysan — APPROVED** [review]

**Pull request merged**